File: hydra-haskell/src/main/haskell/Hydra/Lib/Regex.hs
-- | Haskell implementations of hydra.lib.regex primitives

module Hydra.Lib.Regex where

import qualified Text.Regex.TDFA as R


-- | Check whether an entire string matches a regex pattern.
matches :: String -> String -> Bool
matches pattern input = input R.=~ ("^(" ++ pattern ++ ")$" :: String) :: Bool

-- | Find the first substring matching a regex pattern.
find :: String -> String -> Maybe String
find pattern input = case (input R.=~ pattern :: (String, String, String)) of
  (_, match, _)
    | null match -> Nothing
    | otherwise  -> Just match

-- | Find all non-overlapping substrings matching a regex pattern.
findAll :: String -> String -> [String]
findAll pattern input = R.getAllTextMatches (input R.=~ pattern :: R.AllTextMatches [] String)

-- | Replace the first occurrence of a regex pattern with a replacement string.
replace :: String -> String -> String -> String
replace pattern replacement input = case (input R.=~ pattern :: (String, String, String)) of
  (before, match, after)
    | null match -> input
    | otherwise  -> before ++ replacement ++ after

-- | Replace all non-overlapping occurrences of a regex pattern with a replacement string.
replaceAll :: String -> String -> String -> String
replaceAll pattern replacement input = go input
  where
    go "" = ""
    go s = case (s R.=~ pattern :: (String, String, String)) of
      (before, match, after)
        | null match -> s
        | otherwise  -> before ++ replacement ++ go after

-- | Split a string by a regex pattern.
split :: String -> String -> [String]
split pattern input = go input
  where
    go "" = [""]
    go s = case (s R.=~ pattern :: (String, String, String)) of
      (before, match, after)
        | null match -> [s]
        | otherwise  -> before : go after

File: hydra-python/src/main/python/hydra/lib/regex.py
"""Python implementations of hydra.lib.regex primitives."""

from __future__ import annotations

import re

from hydra.dsl.python import Just, Maybe, NOTHING, frozenlist


def find(pattern: str, input: str) -> Maybe[str]:
    """Find the first substring matching a regex pattern."""
    m = re.search(pattern, input)
    if m is not None:
        return Just(m.group())
    return NOTHING


def find_all(pattern: str, input: str) -> frozenlist[str]:
    """Find all non-overlapping substrings matching a regex pattern."""
    return tuple(re.findall(pattern, input))


def matches(pattern: str, input: str) -> bool:
    """Check whether an entire string matches a regex pattern."""
    return re.fullmatch(pattern, input) is not None


def replace(pattern: str, replacement: str, input: str) -> str:
    """Replace the first occurrence of a regex pattern with a replacement string."""
    return re.sub(pattern, replacement, input, count=1)


def replace_all(pattern: str, replacement: str, input: str) -> str:
    """Replace all non-overlapping occurrences of a regex pattern with a replacement string."""
    return re.sub(pattern, replacement, input)


def split(pattern: str, input: str) -> frozenlist[str]:
    """Split a string by a regex pattern."""
    return tuple(re.split(pattern, input))
